test/tests/functional/pbs_hook_execjob_abort.py

```diff
@@ -0,0 +1,282 @@
+time.sleep(2)
+"""
+        # execjob_prologue hook, unhandled exception
+        self.prolo_hook_body = """import pbs
+e=pbs.event()
+job=e.job
+if %s job.in_ms_mom():
+    x
+"""
+        # execjob_launch hook, unhandled exception on MS mom
+        self.launch_hook_body = """import pbs
+e=pbs.event()
+job=e.job
+if job.in_ms_mom():
+    x
+"""
+        # execjob_end hook
+        self.end_hook_body = """import pbs
+e=pbs.event()
+e.reject("end hook rejected")
+"""
+        # execjob_begin hook with sleep
+        self.begin_hook_body = """import pbs
+import time
+e=pbs.event()
+e.job.delete()
+pbs.logmsg(pbs.LOG_DEBUG, "called execjob_begin hook with job.delete()")
+time.sleep(2)
+"""
+        # job used in the tests
+        a = {ATTR_l + '.select': '3:ncpus=1', ATTR_l + '.place': 'scatter'}
+        self.j = Job(self.du.get_current_user(), attrs=a)
+
+    def test_execjob_abort_ms_prologue(self):
+        """
+        An execjob_abort hook is executed in the primary mom and then in the
+        connected sister moms when a job has problems starting up and needing
+        to be aborted because execjob_prologue hook rejected due to unhandled
+        exception in the primary mom. Job is requeued, gets held (H state).
+        """
+        # instantiate execjob_prologue hook to run on MS mom
+        hook_event = 'execjob_prologue'
+        hook_name = 'prolo'
+        a = {'event': hook_event, 'enabled': 'true'}
+        self.server.create_import_hook(
+            hook_name, a, self.prolo_hook_body % (''))
+        # Submit a job that eventually goes in H state
+        start_time = int(time.time())
+        jid = self.server.submit(self.j)
+        msg = "Job;%s;Job requeued, execution node  down" % jid
+        self.server.log_match(msg, starttime=start_time)
+        # Check for abort hook message in each of the moms
+        msg = "called execjob_abort hook"
+        for mom in self.moms.values():
+            mom.log_match(msg, starttime=start_time)
+        self.server.expect(JOB, {ATTR_state: 'H'}, id=jid, offset=7)
+
+    def test_execjob_abort_exit_job_launch_reject(self):
+        """
+        An execjob_abort hook is executed in the primary mom and then in the
+        connected sister moms when a job has problems starting up and needing
+        to be aborted because execjob_launch hook rejected due to unhandled
+        exception in the primary mom. Job exits.
+        """
+        # instantiate execjob_launch hook to run on primary moms
+        hook_event = 'execjob_launch'
+        hook_name = 'launch'
+        a = {'event': hook_event, 'enabled': 'true'}
+        self.server.create_import_hook(
+            hook_name, a, self.launch_hook_body)
+        # Submit a job
+        start_time = int(time.time())
+        jid = self.server.submit(self.j)
+        self.server.expect(JOB, 'exec_host', id=jid, op=SET)
+        job_stat = self.server.status(JOB, id=jid)
+        exechost = job_stat[0]['exec_host'].partition('/')[0]
```

Comment:
Partition returns a set whereas split returns an ordered list. Either one is probably OK, but sets aren't strictly ordered in Python. I think you should use split instead. (multiple cases)

Reply:
I'll change the multiple instances of 'partition' to 'split'.